core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```
@@ -69,6 +69,11 @@ private[spark] object JsonProtocol extends JsonUtils {
   private[util]
   val defaultOptions: JsonProtocolOptions = new JsonProtocolOptions(new SparkConf(false))
 
+  // SPARK-49872 remove limit on string lengths
+  mapper.getFactory.setStreamReadConstraints(
+    StreamReadConstraints.builder().maxStringLength(Int.MaxValue).build()
```

Two additional limits were introduced in  jackson version 2.15.0. Should we consider setting these to their maximum values as well?

```
      .maxNestingDepth(Int.MaxValue)
      .maxNumberLength(Int.MaxValue)
```

---

Good question. 
I am doubting a little disabling arbitrary limits.   

To be sure, I scanned `JSonProtocol` and I do not see any recursion which can lead to hitting `maxNestingDepth`. 
And I am also not seeing how numbers van have more than 1000 digits in the numbers written bij `JSonProtocol`.

So my advice would be not to do it.  But I am easily convinced otherwise.

---

That's fine with me, thanks for clarifying!

---

also cc @pjfanning

---

`Int.MaxValue` limits the length a string can have to 2 gigabyte.

The reason this option was added to jackson was to give it a more predictable memory usage.  So that in a multi tenant system, you can fail fast when an attacker tries to send you 100 json documents to exhaust your memory and bring down the server.

But in terms of spark history server I think you can argue that spending worst case 2 gigabytes of buffers is acceptable as we do not have that kind of parallel load which can multiply this problem.

But again, I am easily convinced to make it configurable if you feel it is necessary.  But I prefer things to work out of the box.

---

@LuciferYang the change looks ok to me. I would suggest that the Spark Team could consider introducing a new data format instead of storing this data as JSON. Parsing JSON with strings of up to 2Gb will take a lot of heap. Maybe at some stage, Jackson could support a streaming API like SAX where the char data can be broken up into chunks for large strings.

One question: what are you going to do when someone ends up with more than 2Gb of data stored in a big string?

I did notice an issue in JsonUtil. toJsonString wastes a huge amount of memory by not creating a Generator for a StringWriter. Creating on for a ByteArrayOutputSteam and then converting the byte array to a string means you need twice as much memory. You end up with the JSON as byte array and String in memory at same time. Probably is a reasonable amount slower too. Will I create an issue for this?

---

@LuciferYang is there still something preventing this fix for [SPARK-49872](https://github.com/apache/spark/pull/49163#top) to be merged?

Let me know if I can do something.

---

Yeah, this is a restoration of the previous behavior of SHS, so I think it's necessary. 

However, I still hope it can be configurable, and it should have a relatively more reasonable default value. 

Also, could we add a new test case for it? Thanks @steven-aerts 

also cc @dongjoon-hyun FYI

---

@LuciferYang 
I think I see a way to make a test case for it.

About making it configurable, do you have a suggestion for a default value you feel comfortable with?  
We abandoned roczei/spark#7 configurable approach for explicit disabling of this check and go back to the  behavior before jackason 2.15, which is what I am doing here now.

---

To @steven-aerts , I believe @LuciferYang 's suggestion is simply to have a configuration with the default value, `Int.MaxValue`. He also agree your restoration is necessary. So, if you don't mind,
- Please make an internal configuration 
- Also, please provide a unit test case which has a tiny configuration value to make it sure it shows an exception at a slightly bigger string.

WDTY?

---

@LuciferYang @dongjoon-hyun I rewrote the patch and introduced `spark.eventLog.readerMaxStringLength` as an internal configuration parameter which is by default unbounded.
A test case was also added to validate this configuration parameter.
As the spark config now influences the `mapper` inside `JsonProtocol`, I migrated `JsonProtocol` from a singleton object to a class.
I hope you like this more than the previous solution.